src/wmtk/PolygonMesh.cpp

```diff
@@ -0,0 +1,573 @@
+        } else {
+            ++n_halfedges;
+        }
+
+        if (!is_halfedge_connectivity_valid(hid)) {
+            return false;
+        }
+    }
+
+    // Vertex Connectivity
+    // TODO Check vertex boundary conditions
+    long n_vertices = 0;
+    for (long vid = 0; vid < capacity(PrimitiveType::Vertex); ++vid) {
+        // Count vertices
+        if (v_flag_accessor.index_access().scalar_attribute(vid) == 0) {
+            continue;
+        } else {
+            ++n_vertices;
+        }
+
+        // Check local vertex validity
+        if (!is_vertex_connectivity_valid(vid)) {
+            return false;
+        }
+    }
+
+    // Edge Connectivity
+    long n_edges = 0;
+    for (long eid = 0; eid < capacity(PrimitiveType::Edge); ++eid) {
+        // Count edges
+        if (e_flag_accessor.index_access().scalar_attribute(eid) == 0) {
+            continue;
+        } else {
+            ++n_edges;
+        }
+
+        // Check local edge validity
+        if (!is_edge_connectivity_valid(eid)) {
+            return false;
+        }
+    }
+
+    // Face Connectivity
+    long n_faces = 0;
+    for (long fid = 0; fid < capacity(PrimitiveType::Face); ++fid) {
+        // Count faces
+        if (f_flag_accessor.index_access().scalar_attribute(fid) == 0) {
+            continue;
+        } else {
+            ++n_faces;
+        }
+
+        // Check local face validity
+        if (!is_face_connectivity_valid(fid)) {
+            return false;
+        }
+    }
+
+    // Check element counts are consistent
+    if (n_halfedges != (2 * n_edges)) {
+        return false;
+    }
+    if (n_vertices != count_vertices_slow()) {
+        return false;
+    }
+    if (n_faces != count_faces_slow()) {
+        return false;
+    }
+
+    return true;
+}
+
+bool PolygonMesh::is_valid(const Tuple& tuple, ConstAccessor<long>& hash_accessor) const
+{
+    if (tuple.is_null()) return false;
+    if (tuple.m_global_cid < 0) {
+        return false;
+    }
+
+    // Check if local vid is in {0, 1}
+    if ((tuple.m_local_vid != 0) && (tuple.m_local_vid != 1)) {
+        return false;
+    }
+
+    return Mesh::is_hash_valid(tuple, hash_accessor);
+}
+
+bool PolygonMesh::is_ccw(const Tuple& tuple) const
+{
+    return (tuple.m_local_vid == 0);
+}
+
+bool PolygonMesh::is_boundary(const Tuple& tuple) const
+{
+    return is_boundary_edge(tuple);
+}
+
+bool PolygonMesh::is_boundary_vertex(const Tuple& tuple) const
+{
+    Tuple v_tuple = tuple;
+    do {
+        if (is_boundary(v_tuple)) {
+            return true;
+        }
+        v_tuple = switch_edge(switch_face(v_tuple));
+    } while (v_tuple != tuple);
+
+    return false;
+}
+
+bool PolygonMesh::is_boundary_edge(const Tuple& tuple) const
+{
+    return (is_hole_face(tuple)) || (is_hole_face(opp_halfedge(tuple)));
+}
+
+Tuple PolygonMesh::next_halfedge(const Tuple& h_tuple) const
+{
+    return switch_tuple(switch_tuple(h_tuple, PrimitiveType::Vertex), PrimitiveType::Edge);
```

Comment:
there should be a slight contraction to `switch_tuples(h_tuple, {PrimitiveType::Vertex, PrimitiveType::Edge})` which i added a bit ago cuz i think it's easier to read (same might be useful in 217 where you do switch face -> switch edge (i find it very distracting that our two switch notations use opposite typed actions like how reading functions edge(face(tuple)) acts from the left, and then we have ((tuple,face),edge) on the right. if there's more than one switch happening i'd rather use switch_tuples)

Reply:
Can do. The fix for the reversed order would probably be to make the function `switch_tuple(type, tuple)` instead of `switch_tuple(tuple, type)`, right? Go from right to left group action?

Reply:
(Discussed in person)